cmd/juju/commands/main.go

```diff
@@ -406,7 +407,7 @@ func registerCommands(r commandRegistry, ctx *cmd.Context) {
 	r.Register(cloud.NewListCloudsCommand())
 	r.Register(cloud.NewListRegionsCommand())
 	r.Register(cloud.NewShowCloudCommand())
-	r.Register(cloud.NewAddCloudCommand())
+	r.Register(cloud.NewAddCloudCommand(&cloudToCommandAdapter{}))
```

Comment:
Usually what happens is that you begin finding commonality across components, and you can bundle like-concepts together into components and pass them into more than one component. This is one of the main goals/benefits! One example I can immediately think of is there is commonality between the `deploy` and `upgrade` commands. The `deploy` command has been factored to support injection, but `upgrade` has not. As soon as it is, we can factor common code (e.g. resolving charm directives to paths or previously deployed charms) into modules, test those modules once, and then pass them into the commands that need them. You want the root of your call-stack to know more about the system because this allow loser coupling. If you invert that, you get what we have now: tightly coupled components which cannot be modified/tested independently.

If the file-size is an issue, we can always split it out into more files.